azkaban-common/src/main/java/azkaban/project/FlowTrigger.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2017 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.project;
+
+import azkaban.Constants;
+import azkaban.utils.Props;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * FlowTrigger is an immutable class which holds
+ * all the data and properties of a flow trigger.
+ */
+
+public class FlowTrigger {
+
+  private final List<FlowTriggerDependency> dependencies;
+  private final Props props; // trigger level props
```

Comment:
Is it better to stored typed value here too? 
It's not clear what properties are allowed here.

Reply:
e.g. if the caller makes a mistake, passing a prop without the name, the constructor will throw an exception. If you define the method that explicitly requires the needed information, the compiler will do the checking for you and fail at the compile time.
You won't need the following validation code here.
You will need the validation code in the parsing logic. And even there, the compiler will remind that author what information is required.

   final String MISSING_REQUIRED_ERROR = "missing required param: %s";
 +
 +    final Set<String> requiredParam = ImmutableSet.of(Constants.TriggerProperties.SCHEDULE_TYPE,
 +        Constants.TriggerProperties.SCHEDULE_VALUE,
 +        Constants.TriggerProperties.SCHEDULE_MAX_WAIT_TIME);
 +
 +    for (final String param : requiredParam) {
 +      Preconditions.checkArgument(props.containsKey(param), String.format(MISSING_REQUIRED_ERROR,
 +          param));
 +    }

Reply:
You don't seem to have tests for missing a required information from the props. 

Could you check your test coverage?

Reply:
updated